File: backend/src/main/java/com/autocare/inventory/controller/InventoryController.java
package com.autocare.inventory.controller;

import com.autocare.inventory.dto.InventoryMovementDTO;
import com.autocare.inventory.dto.PartRequestDTO;
import com.autocare.inventory.dto.PartResponseDTO;
import com.autocare.inventory.service.InventoryService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/inventory")
@RequiredArgsConstructor
public class InventoryController {

    private final InventoryService inventoryService;

    // Parts
    @PostMapping("/parts")
    public ResponseEntity<PartResponseDTO> createPart(@Valid @RequestBody PartRequestDTO request) {
        return ResponseEntity.status(HttpStatus.CREATED).body(inventoryService.createPart(request));
    }

    @GetMapping("/parts")
    public ResponseEntity<List<PartResponseDTO>> findAllParts() {
        return ResponseEntity.ok(inventoryService.findAllParts());
    }

    @GetMapping("/parts/low-stock")
    public ResponseEntity<List<PartResponseDTO>> findLowStockParts() {
        return ResponseEntity.ok(inventoryService.findLowStockParts());
    }

    @GetMapping("/parts/{id}")
    public ResponseEntity<PartResponseDTO> findPartById(@PathVariable String id) {
        return ResponseEntity.ok(inventoryService.findPartByIdResponse(id));
    }

    @PutMapping("/parts/{id}")
    public ResponseEntity<PartResponseDTO> updatePart(
            @PathVariable String id,
            @Valid @RequestBody PartRequestDTO request) {
        return ResponseEntity.ok(inventoryService.updatePart(id, request));
    }

    @DeleteMapping("/parts/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public ResponseEntity<Void> deletePart(@PathVariable String id) {
        inventoryService.deletePart(id);
        return ResponseEntity.noContent().build();
    }

    // Stock Movements
    @PostMapping("/parts/{id}/add-stock")
    public ResponseEntity<Void> addStock(
            @PathVariable String id,
            @RequestParam Integer quantity,
            @RequestParam String reason,
            @RequestParam String userId) {
        inventoryService.addStock(id, quantity, reason, userId);
        return ResponseEntity.noContent().build();
    }

    @PostMapping("/parts/{id}/remove-stock")
    public ResponseEntity<Void> removeStock(
            @PathVariable String id,
            @RequestParam Integer quantity,
            @RequestParam String reason,
            @RequestParam String userId) {
        inventoryService.removeStock(id, quantity, reason, userId);
        return ResponseEntity.noContent().build();
    }

    @GetMapping("/parts/{id}/movements")
    public ResponseEntity<List<InventoryMovementDTO>> getPartMovements(@PathVariable String id) {
        return ResponseEntity.ok(inventoryService.getPartMovements(id));
    }
}

File: backend/src/inventory/inventory.controller.ts
import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  ParseIntPipe,
  Post,
  Put,
  Query,
  UseGuards,
} from "@nestjs/common";
import { InventoryService } from "./inventory.service";
import { PartRequestDto } from "./dto/part.dto";
import { Roles } from "../common/decorators/roles.decorator";
import { RolesGuard } from "../common/guards/roles.guard";
import { CurrentUser } from "../common/decorators/current-user.decorator";
import { Role } from "@prisma/client";

// Conforme a tabela de permissões do projeto, o módulo de estoque é restrito a
// ADMIN e MANAGER (RECEPTIONIST e MECHANIC não devem criar peças ou movimentar
// estoque diretamente).
@Controller("inventory")
@UseGuards(RolesGuard)
@Roles(Role.ADMIN, Role.MANAGER)
export class InventoryController {
  constructor(private readonly inventoryService: InventoryService) {}

  @Post("parts")
  @HttpCode(HttpStatus.CREATED)
  createPart(@Body() dto: PartRequestDto) {
    return this.inventoryService.createPart(dto);
  }

  @Get("parts")
  findAllParts() {
    return this.inventoryService.findAllParts();
  }

  @Get("parts/low-stock")
  findLowStockParts() {
    return this.inventoryService.findLowStockParts();
  }

  @Get("parts/:id")
  findPartById(@Param("id") id: string) {
    return this.inventoryService.findPartByIdResponse(id);
  }

  @Put("parts/:id")
  updatePart(@Param("id") id: string, @Body() dto: PartRequestDto) {
    return this.inventoryService.updatePart(id, dto);
  }

  @Delete("parts/:id")
  @HttpCode(HttpStatus.NO_CONTENT)
  async deletePart(@Param("id") id: string) {
    await this.inventoryService.deletePart(id);
  }

  // O userId de auditoria vem sempre do usuário autenticado (JWT via @CurrentUser),
  // nunca de um parâmetro enviado pelo cliente — antes era possível forjar esse valor
  // e registrar uma movimentação de estoque em nome de outra pessoa.
  @Post("parts/:id/add-stock")
  @HttpCode(HttpStatus.NO_CONTENT)
  async addStock(
    @Param("id") id: string,
    @Query("quantity", ParseIntPipe) quantity: number,
    @Query("reason") reason: string,
    @CurrentUser("id") userId: string,
  ) {
    await this.inventoryService.addStock(id, quantity, reason, userId);
  }

  @Post("parts/:id/remove-stock")
  @HttpCode(HttpStatus.NO_CONTENT)
  async removeStock(
    @Param("id") id: string,
    @Query("quantity", ParseIntPipe) quantity: number,
    @Query("reason") reason: string,
    @CurrentUser("id") userId: string,
  ) {
    await this.inventoryService.removeStock(id, quantity, reason, userId);
  }

  @Get("parts/:id/movements")
  getPartMovements(@Param("id") id: string) {
    return this.inventoryService.getPartMovements(id);
  }
}
